File: impl/go/certverify_kat_test.go
package npamp

import (
	"bytes"
	"crypto/ed25519"
	"errors"
	"testing"
)

// certVerifyKAT mirrors test-vectors/v1/certverify-kat.json (ADR-0011):
// CertVerify value = u16(0x0807) || Ed25519(priv, signing_input), with
// signing_input = 0x20 x 64 || context || 0x00 || transcript_hash (spec/10
// section 6.1). The Ed25519 primitive is anchored to RFC 8032 section 7.1
// TEST 1/TEST 2; the keys are the RFC 8032 test keypairs; the transcript
// hashes are the Transcript KAT's TH_sId / TH_cId.
type certVerifyKAT struct {
	Contexts struct {
		Client string `json:"client"`
		Server string `json:"server"`
	} `json:"contexts"`
	Expected struct {
		ValueClient  string `json:"certverify_value_client"`
		ValueServer  string `json:"certverify_value_server"`
		SigClient    string `json:"signature_client"`
		SigServer    string `json:"signature_server"`
		InputClient  string `json:"signing_input_client"`
		InputServer  string `json:"signing_input_server"`
	} `json:"expected"`
	Inputs struct {
		ClientPub  string `json:"client_pub"`
		ClientSeed string `json:"client_seed"`
		ServerPub  string `json:"server_pub"`
		ServerSeed string `json:"server_seed"`
		THCID      string `json:"th_cid"`
		THSID      string `json:"th_sid"`
	} `json:"npamp_inputs"`
	RFC8032 struct {
		Test1 struct {
			Message   string `json:"message"`
			PublicKey string `json:"public_key"`
			Seed      string `json:"seed"`
			Signature string `json:"signature"`
		} `json:"test1"`
		Test2 struct {
			Message   string `json:"message"`
			PublicKey string `json:"public_key"`
			Seed      string `json:"seed"`
			Signature string `json:"signature"`
		} `json:"test2"`
	} `json:"rfc8032_ed25519"`
}

func TestCertVerifyKAT(t *testing.T) {
	var kat certVerifyKAT
	loadKAT(t, "certverify-kat.json", &kat)

	// Anchor (RFC 8032 section 7.1 TEST 1/TEST 2): prove the Ed25519
	// primitive — public-key derivation and the deterministic signature.
	for name, tv := range map[string]struct{ Message, PublicKey, Seed, Signature string }{
		"test1": {kat.RFC8032.Test1.Message, kat.RFC8032.Test1.PublicKey, kat.RFC8032.Test1.Seed, kat.RFC8032.Test1.Signature},
		"test2": {kat.RFC8032.Test2.Message, kat.RFC8032.Test2.PublicKey, kat.RFC8032.Test2.Seed, kat.RFC8032.Test2.Signature},
	} {
		priv := ed25519.NewKeyFromSeed(mustHex(t, name+".seed", tv.Seed))
		if !bytes.Equal(priv.Public().(ed25519.PublicKey), mustHex(t, name+".public_key", tv.PublicKey)) {
			t.Fatalf("Ed25519 public key does not reproduce RFC 8032 %s", name)
		}
		sig := ed25519.Sign(priv, mustHex(t, name+".message", tv.Message))
		if !bytes.Equal(sig, mustHex(t, name+".signature", tv.Signature)) {
			t.Fatalf("Ed25519 signature does not reproduce RFC 8032 %s", name)
		}
	}

	serverPriv := ed25519.NewKeyFromSeed(mustHex(t, "server_seed", kat.Inputs.ServerSeed))
	clientPriv := ed25519.NewKeyFromSeed(mustHex(t, "client_seed", kat.Inputs.ClientSeed))
	serverPub := mustHex(t, "server_pub", kat.Inputs.ServerPub)
	clientPub := mustHex(t, "client_pub", kat.Inputs.ClientPub)
	thSID := mustHex(t, "th_sid", kat.Inputs.THSID)
	thCID := mustHex(t, "th_cid", kat.Inputs.THCID)

	cases := []struct {
		roleName  string
		role      Role
		priv      ed25519.PrivateKey
		pub       []byte
		th        []byte
		context   string
		wantInput []byte
		wantSig   []byte
		wantValue []byte
	}{
		{"server", RoleServer, serverPriv, serverPub, thSID, kat.Contexts.Server,
			mustHex(t, "signing_input_server", kat.Expected.InputServer),
			mustHex(t, "signature_server", kat.Expected.SigServer),
			mustHex(t, "certverify_value_server", kat.Expected.ValueServer)},
		{"client", RoleClient, clientPriv, clientPub, thCID, kat.Contexts.Client,
			mustHex(t, "signing_input_client", kat.Expected.InputClient),
			mustHex(t, "signature_client", kat.Expected.SigClient),
			mustHex(t, "certverify_value_client", kat.Expected.ValueClient)},
	}
	for _, c := range cases {
		// Oracle leg: build the signing input independently of the impl —
		// 64 x 0x20 || context || 0x00 || transcript_hash — and sign with
		// crypto/ed25519 directly.
		oracleInput := append(bytes.Repeat([]byte{0x20}, 64), []byte(c.context)...)
		oracleInput = append(oracleInput, 0x00)
		oracleInput = append(oracleInput, c.th...)
		if !bytes.Equal(oracleInput, c.wantInput) {
			t.Fatalf("%s: oracle signing input does not reproduce the vector", c.roleName)
		}
		if got := ed25519.Sign(c.priv, oracleInput); !bytes.Equal(got, c.wantSig) {
			t.Fatalf("%s: oracle Ed25519 signature does not reproduce the vector", c.roleName)
		}
		if got := concat([]byte{0x08, 0x07}, c.wantSig); !bytes.Equal(got, c.wantValue) {
			t.Fatalf("%s: certverify_value is not u16(0x0807) || signature", c.roleName)
		}

		// Impl leg: the real signing-input builder, signer, and verifier.
		implInput, err := CertVerifySigningInput(c.role, c.th)
		if err != nil {
			t.Fatal(err)
		}
		if !bytes.Equal(implInput, c.wantInput) {
			t.Fatalf("%s: CertVerifySigningInput = %x, want %x", c.roleName, implInput, c.wantInput)
		}
		value, err := SignCertVerify(c.priv, c.role, c.th)
		if err != nil {
			t.Fatal(err)
		}
		if !bytes.Equal(value, c.wantValue) {
			t.Fatalf("%s: SignCertVerify = %x, want %x", c.roleName, value, c.wantValue)
		}
		if err := VerifyCertVerify(c.pub, c.role, c.th, value); err != nil {
			t.Fatalf("%s: VerifyCertVerify rejected the correct value: %v", c.roleName, err)
		}

		// Domain separation: the same value presented under the OTHER role's
		// context MUST be rejected (spec/10 section 6.1).
		otherRole := RoleClient
		if c.role == RoleClient {
			otherRole = RoleServer
		}
		if err := VerifyCertVerify(c.pub, otherRole, c.th, value); !errors.Is(err, ErrCertVerifySignature) {
			t.Fatalf("%s: role-mismatched CertVerify not rejected (err=%v)", c.roleName, err)
		}

		// Transcript binding: a different transcript hash MUST be rejected.
		otherTH := thCID
		if c.role == RoleClient {
			otherTH = thSID
		}
		if err := VerifyCertVerify(c.pub, c.role, otherTH, value); !errors.Is(err, ErrCertVerifySignature) {
			t.Fatalf("%s: wrong-transcript CertVerify not rejected (err=%v)", c.roleName, err)
		}

		// Scheme pinning: a non-negotiated SignatureScheme MUST be rejected.
		badScheme := bytes.Clone(value)
		badScheme[0], badScheme[1] = 0x09, 0x05 // ML-DSA-87 code point, not negotiated at Standard
		if err := VerifyCertVerify(c.pub, c.role, c.th, badScheme); !errors.Is(err, ErrCertVerifyScheme) {
			t.Fatalf("%s: non-negotiated scheme not rejected (err=%v)", c.roleName, err)
		}
	}
}

File: impl/typescript/test/certverify-kat.test.ts
// Standards-derived, NON-CIRCULAR known-answer test for the draft-00 CertVerify (binding spec/10
// §6.1; RFC 8446 §4.4.3 structure; Ed25519 per RFC 8032). The value is u16(0x0807) ||
// Ed25519(priv, signing_input), signing_input = 64*0x20 || context || 0x00 || TH. TypeScript mirror
// of the Go reference test against the SAME pinned vector.
//
// NON-CIRCULARITY: the expected signatures in test-vectors/v1/certverify-kat.json were produced with
// Ed25519 directly, NOT with signCertVerify/certVerifySigningInput. Three legs:
//  1. ANCHOR — the src Ed25519 key helpers reproduce the published RFC 8032 §7.1 TEST 1/TEST 2
//     public keys + signatures (trust the primitive + the PKCS8/SPKI wiring).
//  2. ORACLE — rebuild signing_input by hand and re-sign with an independently-constructed key,
//     reproducing the vector (guards the vector; no src signing functions).
//  3. IMPL   — certVerifySigningInput + signCertVerify reproduce the vector; verifyCertVerify
//     accepts the correct value but REJECTS a role/context mismatch and a wrong transcript (guards
//     the impl + the domain-separation property).
import { test } from "node:test";
import assert from "node:assert";
import { createHash, createPrivateKey, createPublicKey, sign, verify } from "node:crypto";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import {
  certVerifySigningInput, signCertVerify, verifyCertVerify,
  ed25519PrivateKeyFromSeed, ed25519PublicKeyFromRaw,
  CONTEXT_SERVER_CERTVERIFY, CONTEXT_CLIENT_CERTVERIFY,
} from "../src/npamp.ts";

const VECTORS = join(import.meta.dirname, "..", "..", "..", "test-vectors", "v1");
const CERTVERIFY_KAT_SHA256 = "f56ec6ba250ba8f8c6c84214a16f580a3e476e9b2cfd05720c3352de299fe555";

const hx = (s) => Buffer.from(s, "hex");

function loadCertVerifyKAT() {
  const raw = readFileSync(join(VECTORS, "certverify-kat.json"));
  const got = createHash("sha256").update(raw).digest("hex");
  assert.equal(got, CERTVERIFY_KAT_SHA256, "CertVerify KAT vector SHA-256 mismatch (swapped vector?)");
  return JSON.parse(raw.toString("utf8"));
}

// rawPub extracts the 32-octet raw Ed25519 public key from a KeyObject (via JWK x).
function rawPub(keyObject) {
  return Buffer.from(keyObject.export({ format: "jwk" }).x, "base64url");
}

// Oracle key construction (RFC 8410 PKCS8 wrap), independent of the src helpers.
const PKCS8_ED25519_PREFIX = Buffer.from("302e020100300506032b657004220420", "hex");
function oraclePriv(seed) {
  return createPrivateKey({ key: Buffer.concat([PKCS8_ED25519_PREFIX, seed]), format: "der", type: "pkcs8" });
}
// Oracle signing_input, built by hand (independent of certVerifySigningInput).
function oracleSigningInput(ctx, th) {
  return Buffer.concat([Buffer.alloc(64, 0x20), Buffer.from(ctx), Buffer.from([0x00]), th]);
}

// ANCHOR: the src Ed25519 helpers reproduce the published RFC 8032 §7.1 keys + signatures.
test("certverify_kat_rfc8032_anchor", () => {
  const k = loadCertVerifyKAT();
  for (const [name, v] of [["TEST1", k.rfc8032_ed25519.test1], ["TEST2", k.rfc8032_ed25519.test2]]) {
    const priv = ed25519PrivateKeyFromSeed(hx(v.seed));
    assert.equal(rawPub(createPublicKey(priv)).toString("hex"), v.public_key, `${name} derived pubkey != RFC 8032`);
    const sig = sign(null, hx(v.message), priv);
    assert.equal(sig.toString("hex"), v.signature, `${name} signature != RFC 8032`);
    const pub = ed25519PublicKeyFromRaw(hx(v.public_key));
    assert.ok(verify(null, hx(v.message), pub, hx(v.signature)), `${name} ed25519PublicKeyFromRaw failed to verify`);
  }
});

// ORACLE: rebuild signing_input by hand and re-sign with an independent key, reproducing the vector.
test("certverify_kat_oracle", () => {
  const k = loadCertVerifyKAT();
  const n = k.npamp_inputs, e = k.expected, c = k.contexts;
  for (const role of [
    { name: "server", ctx: c.server, seed: n.server_seed, th: n.th_sid, wantSI: e.signing_input_server, wantSig: e.signature_server },
    { name: "client", ctx: c.client, seed: n.client_seed, th: n.th_cid, wantSI: e.signing_input_client, wantSig: e.signature_client },
  ]) {
    const si = oracleSigningInput(role.ctx, hx(role.th));
    assert.equal(si.toString("hex"), role.wantSI, `[${role.name}] oracle signing_input != vector`);
    const sig = sign(null, si, oraclePriv(hx(role.seed)));
    assert.equal(sig.toString("hex"), role.wantSig, `[${role.name}] oracle signature != vector`);
  }
});

// IMPL: certVerifySigningInput + signCertVerify reproduce the vector; verifyCertVerify accepts the
// correct value, rejects a role/context mismatch and a wrong transcript.
test("certverify_kat_impl", () => {
  const k = loadCertVerifyKAT();
  const n = k.npamp_inputs, e = k.expected, c = k.contexts;
  assert.equal(CONTEXT_SERVER_CERTVERIFY, c.server, "server CertVerify context constant drifted from spec §6.1");
  assert.equal(CONTEXT_CLIENT_CERTVERIFY, c.client, "client CertVerify context constant drifted from spec §6.1");
  for (const role of [
    { name: "server", isServer: true, seed: n.server_seed, pub: n.server_pub, th: n.th_sid, wantSI: e.signing_input_server, wantValue: e.certverify_value_server },
    { name: "client", isServer: false, seed: n.client_seed, pub: n.client_pub, th: n.th_cid, wantSI: e.signing_input_client, wantValue: e.certverify_value_client },
  ]) {
    const priv = ed25519PrivateKeyFromSeed(hx(role.seed));
    const pub = ed25519PublicKeyFromRaw(hx(role.pub));
    const th = hx(role.th);

    assert.equal(certVerifySigningInput(role.isServer, th).toString("hex"), role.wantSI, `[${role.name}] certVerifySigningInput != vector`);

    const val = signCertVerify(priv, role.isServer, th);
    assert.equal(val.toString("hex"), role.wantValue, `[${role.name}] signCertVerify value != vector`);

    assert.ok(verifyCertVerify(pub, role.isServer, th, val), `[${role.name}] verifyCertVerify rejected the correct value`);
    // Domain separation: verifying with the opposite role must FAIL (different context string).
    assert.ok(!verifyCertVerify(pub, !role.isServer, th, val), `[${role.name}] verifyCertVerify accepted a role/context mismatch`);
    // Transcript binding: a different transcript hash must FAIL.
    const wrongTH = Buffer.from(th);
    wrongTH[0] ^= 0x01;
    assert.ok(!verifyCertVerify(pub, role.isServer, wrongTH, val), `[${role.name}] verifyCertVerify accepted a wrong transcript hash`);
    // Scheme guard: a non-Ed25519 scheme code point must FAIL.
    const wrongScheme = Buffer.from(val);
    wrongScheme.writeUInt16BE(0x0905, 0); // ML-DSA-87 code point, not Ed25519 (0x0807)
    assert.ok(!verifyCertVerify(pub, role.isServer, th, wrongScheme), `[${role.name}] verifyCertVerify accepted a non-Ed25519 scheme`);
    // Length guard: an Ed25519 signature is exactly 64 octets; a truncated value must FAIL.
    assert.ok(!verifyCertVerify(pub, role.isServer, th, val.subarray(0, val.length - 1)), `[${role.name}] verifyCertVerify accepted a truncated signature`);
  }
});
